src/Tinker.jl

```diff
@@ -1,4 +1,268 @@
+    end
+
+    # holds x zoom level
+    xzoom = map(zr) do r
+        100*r.fullview.x.right/(r.currentview.x.right-(r.currentview.x.left-1))
+    end
+
+    # hold y zoom level
+    yzoom = map(zr) do r
+        100*r.fullview.y.right/(r.currentview.y.right-(r.currentview.y.left-1))
+    end
+
+    zpercents = [1.0,1.2,1.5,2.0,2.5,3.0,4.0,8.0]
+    global i = 1 # or: const i = Ref(1)
+
+    # performs proportional zoom in
+    function zoomin(center::XY{Int})
+        global i
+        if 1 <= i <= length(zpercents)
+            if i < length(zpercents)
+                i += 1
+                push!(zr, zoom_percent(zpercents[i],value(zr),center))
+            end
+        else
+            index = 1
+            for n in zpercents
+                if n > value(xzoom)
+                    break
+                end
+                index += 1
+            end
+            i = index
+            push!(zr, zoom_percent(zpercents[i],value(zr),center))
+        end
+    end
+
+    function zoomin()
+        global i
+        if 1 <= i <= length(zpercents)
+            if i < length(zpercents)
+                i += 1
+                push!(zr, zoom_percent(zpercents[i],value(zr)))
+            end
+        else
+            index = 1
+            for n in zpercents
+                if n > value(xzoom)
+                    break
+                end
+                index += 1
+            end
+            i = index
+            push!(zr, zoom_percent(zpercents[i],value(zr)))
+        end
+    end
+
+    # performs proportional zoom out
+    function zoomout(center::XY{Int})
+        global i
+        if 1 <= i <= length(zpercents)
+            if i > 1
+                i -= 1
+                push!(zr, zoom_percent(zpercents[i],value(zr),center))
+            end
+        else
+            index = length(zpercents)
+            for n in zpercents[end:-1:1] # loop backwards      
+                if n < value(xzoom) 
+                    break
+                end
+                index -= 1
+            end
+            i = index
+            push!(zr, zoom_percent(zpercents[i],value(zr),center))
+        end
+    end
+
+    function zoomout()
+        global i
+        if 1 <= i <= length(zpercents)
+            if i > 1
+                i -= 1
+                push!(zr, zoom_percent(zpercents[i],value(zr)))
+            end
+        else
+            index = length(zpercents)
+            for n in zpercents[end:-1:1] # loop backwards      
+                if n < value(xzoom) 
+                    break
+                end
+                index -= 1
+            end
+            i = index
+            push!(zr, zoom_percent(zpercents[i],value(zr)))
+        end
+    end
+
+    # performs proportional, centered zoom to level entered
+    function zoom_to(z::Float64)
+        global i
+        push!(zr, zoom_percent(z,value(zr)))
+        i = -1
+        nothing
+    end
+
+    showall(win);
+
+    function zoom_clicked{T}(c::GtkReactive.Canvas,
```

Comment:
At some point I predict you're going to want to support different mouse-modes (e.g., when zooming vs drawing on the image), but that should be a subject for a future PR. More `filterwhen` statements seem likely to be the solution.